internal/command/sandbox/subst.go

```diff
@@ -0,0 +1,182 @@
+package sandbox
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/signadot/cli/internal/clio"
+	"github.com/signadot/cli/internal/config"
+	"github.com/signadot/go-sdk/models"
+)
+
+func loadSandbox(file string, tplVals config.TemplateVals) (*models.Sandbox, error) {
+	substMap, err := substMap(tplVals)
+	if err != nil {
+		return nil, err
+	}
+	sbt, err := clio.LoadYAML[any](file)
+	if err != nil {
+		return nil, err
+	}
+	if err := substTemplate(sbt, substMap); err != nil {
+		return nil, err
+	}
+	return unstructuredToSandbox(*sbt)
+}
+
+func substMap(tplVals []config.TemplateVal) (map[string]string, error) {
+	substMap := map[string]string{}
+	conflicts := map[string][]string{}
+	for _, tv := range tplVals {
+		if tVal, present := substMap[tv.Var]; present {
+			if tVal != tv.Val {
+				if len(conflicts[tv.Var]) == 0 {
+					conflicts[tv.Var] = []string{tv.Val}
+				}
+				conflicts[tv.Var] = append(conflicts[tv.Var], tVal)
+				continue
+			}
+		}
+		substMap[tv.Var] = tv.Val
+	}
+	if len(conflicts) == 0 {
+		return substMap, nil
+	}
+	conflictKeys := make([]string, 0, len(conflicts))
+	for k := range conflicts {
+		conflictKeys = append(conflictKeys, k)
+	}
+	sort.Strings(conflictKeys)
+	msgs := make([]string, 0, len(conflictKeys))
+	for _, key := range conflictKeys {
+		vals := strings.Join(conflicts[key], ", ")
+		msgs = append(msgs, fmt.Sprintf("%s: %s", key, vals))
+	}
+	return nil, fmt.Errorf("conflicting variable defs:\n\t%s\n", strings.Join(msgs, "\n\t"))
+}
+
+func substTemplate(sbt *any, substMap map[string]string) error {
+	vars := map[string]struct{}{}
+	err := substTemplateRec(sbt, substMap, vars)
+	if err != nil {
+		return err
+	}
+	notExpanded := []string{}
+	for k := range vars {
+		if _, ok := substMap[k]; !ok {
+			notExpanded = append(notExpanded, k)
+		}
+	}
+	if len(notExpanded) > 0 {
+		return fmt.Errorf("unexpanded variables: %s", strings.Join(notExpanded, ", "))
+	}
+	return nil
+}
+
+func substTemplateRec(sbt *any, substMap map[string]string, vars map[string]struct{}) error {
+	switch x := (*sbt).(type) {
+	case map[string]any:
+		for k, v := range x {
+			if err := substTemplateRec(&v, substMap, vars); err != nil {
+				return err
+			}
+			x[k] = v
+		}
+
+	case []any:
+		for _, v := range x {
+			if err := substTemplateRec(&v, substMap, vars); err != nil {
```

Comment:
It works on tests; I think the reason is that v is type `any`, and so has a pointer built-in to it.

I do not think we need `x[I]= v` because it is updated by pointer for case string.

In any event, to make it more readable, I've replaced v with &x[I] in next commit.